### PR TITLE
In-memory Engine: optimize write batch to selectively cache writes

### PR DESCRIPTION
**Pull request by SpadeA-Tang**
### What is changed and how it works?

Issue Number: Ref #16141

What's Changed:

```commit-message
optimize write batch to selectively cache writes
```
The current write batch buffers all writes for in-memory engine even the ranges of the writes are not cached. This PR optimize it to only buffer writes whose range is qualified (the normal range or pending range that is loading data).

### Related changes

- [ ] PR to update `pingcap/docs`/`pingcap/docs-cn`:
- [ ] Need to cherry-pick to the release branch

### Check List

Tests 

- [x] Unit test
- [x] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: components/region_cache_memory_engine/src/write_batch.rs [hunk 1]
```diff
@@ -1,20 +1,29 @@
...
 use crate::{
-    engine::{cf_to_id, RangeCacheMemoryEngineCore, SkiplistEngine},
+    engine::{cf_to_id, SkiplistEngine},
     keys::{encode_key, ValueType},
     range_manager::RangeManager,
     RangeCacheMemoryEngine,
 };
 
+// `prepare_for_range` should be called before raft command apply for each peer
+// delete. It sets `range_in_cache` and `pending_range_in_loading` which is used
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/write_batch.rs:15]: delete -> delegate

File: components/region_cache_memory_engine/src/write_batch.rs [hunk 2]
```diff
@@ -279,8 +318,13 @@ impl Mutable for RangeCacheWriteBatch {
...
     fn put_cf(&mut self, cf: &str, key: &[u8], val: &[u8]) -> Result<()> {
-        self.buffer
-            .push(RangeCacheWriteBatchEntry::put_value(cf, key, val));
+        if self.range_in_cache {
+            self.buffer
+                .push(RangeCacheWriteBatchEntry::put_value(cf, key, val));
+        } else if self.pending_range_in_loading {
+            self.pending_range_in_loading_buffer
+                .push(RangeCacheWriteBatchEntry::put_value(cf, key, val));
+        }
```
Comment (tonyxuqqi) [inline components/region_cache_memory_engine/src/write_batch.rs:327]: Can the code of put_cf and delete_cf be shared on this code path? 
This is critical part that missing any entry could lead to data loss.

**Review by tonyxuqqi — APPROVED** [review]
Overall LGTM.

File: components/region_cache_memory_engine/src/write_batch.rs (hunk 2 above)
Comment (SpadeA-Tang) [reply components/region_cache_memory_engine/src/write_batch.rs]: Done

File: components/engine_traits/src/write_batch.rs [hunk 3]
```diff
@@ -123,4 +123,8 @@ pub trait WriteBatch: Mutable {
 
     /// Merge another WriteBatch to itself
     fn merge(&mut self, src: Self) -> Result<()>;
+
+    /// It declears that the following consecutive write will be within this
```
Comment (afeinberg) [inline components/engine_traits/src/write_batch.rs:127]:
```suggestion
    /// It declares that the following consecutive write will be within this
```

**Review by afeinberg — APPROVED** [review]
LGTM.

**Review by LykxSassinator — APPROVED** [review]

**Pull request merged**